docs/proposals/20210222-kubelet-authentication.md

```diff
@@ -0,0 +1,319 @@
+    - cluster.x-k8s.io/kube-apiserver-client-kubelet-tpm and cluster-x-k8s-io/kubelet-serving-tpm
+        - Will implement TPM-based certificate signers and requesters based on the cloud-provider-gcp implementation.
+        - However, since the mechanism for retrieving endorsement keys varies across platforms, the TPM signer will additionally require a provider specific mechanism to provide the TPM Endorsement Key.
+
+#### Provider Specification
+
+#### All providers
+- All providers must insert a ProviderID within the KubernetesNodeProviderIdentifierOID extension attribute of the CSR
+- All signer names must be filled in by the provider’s controller in InfraCluster.Status.KubeletClientCertificateSigner and InfraCluster.Status.KubeletServingCertificateSigner
+
+#### Insecure providers
+- An insecure provider MUST not implement certificate rotation or kubelet serving certificate signing
+- InfraCluster.Status.KubeletClientCertificateSigner MUST be set to cluster.x-k8s.io/kube-apiserver-client-kubelet-insecure
+- An insecure provider MUST use the cluster.x-k8s.io/kube-apiserver-client-kubelet-insecure signer/v1alpha4
+
+#### Secure providers
+- A secure provider MUST implement certificate rotation and kubelet server certificate signing
+- A provider must register signers of:
+    - `cluster-x-k8s-io/kube-apiserver-client-kubelet-<provider>/v1alpha4`
+    - `cluster-x-k8s-io/kubelet-serving-<provider>/v1alpha4`
+- A secure provider MUST implement a secure attestation mechanism, based upon PEM-encoded blocks within the Certificate Signing Request.
+- Where a secure provider’s attestation mechanism does not include a nonce or timestamp to protect against replay attacks, the mechanism MUST implement a secondary time-limited attestation (e.g. AWS Instance Identity document + AWS HMACv4 signature).
+- A provider’s signer MUST run on the management cluster
+
+#### TPM based providers
+- A TPM provider MUST use the following certificate signers
+    - `cluster-x-k8s-io/kube-apiserver-client-kubelet-tpm/v1alpha4`
+    - `cluster-x-k8s-io/kubelet-serving-tpm/v1alpha4`
+- A TPM provider MUST annotate new CSRs as follows:
+    - Key: cluster-x-k8s-io/tpm-endorsement-key/v1alpha4
+    - Value: Platform-specific endorsement key (e.g., retrieved from GCP Shielded VM API or VMware vCenter)
+
+#### Kubeadm
+Since this proposal essentially takes over part of the node registration process from kubeadm, we will require the following changes:
+- kubeadm COULD allow opt-out of kubeadm setting up ClusterRoleBindings between the system:nodes group and the `system:certificates.k8s.io:certificatesigningrequests:selfnodeclient` permission, so that certificate renewals must go through re-attestation. However, Cluster API kubeadm control plane provider is able to reconcile the removal of these roles during control plane instantiation.
+- Kubeadm COULD allow opt-out of kubeadm setting up the bootstrapper and node groups and bindings.
```

Comment:
could you elaborate on what does this entail?

Reply:
Do we need details here, or can we defer to a Kubeadm KEP for v1beta3?

Reply:
we already discussed what is planned for v1beta3 and established some priorities.
i would prefer to see some details here, even as a comment, before we jump into a KEP.

Reply:
Updated